### PR TITLE
`azurerm_app_configuration_key`, `azurerm_app_configuration_feature` - change resource ID to Data Plane URL format

### LINKED ISSUE #20038: azurerm_app_configuration_key - while parsing resource ID: the number of path segments is not divisible by 2
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

1.2.5

### AzureRM Provider Version

3.39.1

### Affected Resource(s)/Data Source(s)

azurerm_app_configuration_key

### Terraform Configuration Files

```hcl
resource "azurerm_app_configuration_key" "this" {
  for_each               = fileset("${path.cwd}/configurations/${terraform.workspace}", "*")
  configuration_store_id = var.app_config_id
  key                    = upper(split(".", each.key)[0])
  type                   = "kv"
  value                  = file("${path.cwd}/configurations/${terraform.workspace}/${each.value}")
}
```

### Debug Output/Panic Output

```shell
Error: while parsing resource ID: while parsing resource ID: the number of path segments is not divisible by 2 in "subscriptions/[MASKED]/resourceGroups/[MASKED]providers/Microsoft.AppConfiguration/configurationStores/[MASKED]AppConfigurationKey/SAMPLE%2FLabel%2F%00%2FAppConfigurationKey%2FSAMPLE/Label"
│ 
│   with azurerm_app_configuration_key.this["sample.json"],
│   on main.tf line 35, in resource "azurerm_app_configuration_key" "this":
│   35: resource "azurerm_app_configuration_key" "this" {
│ 
│ while parsing resource ID: while parsing resource ID: the number of path
│ segments is not divisible by 2 in
│ "subscriptions/[MASKED]/resourceGroups/[MASKED]providers/Microsoft.AppConfiguration/configurationStores/[MASKED]AppConfigurationKey/SAMPLE%2FLabel%2F%00%2FAppConfigurationKey%2FSAMPLE/Label"
```

### Expected Behaviour

The provider should be able to detect the changes in the sample.json file and update the key in App Configuration.

### Actual Behaviour

The terraform plan fails with the following error

`Error: while parsing resource ID: while parsing resource ID: the number of path segments is not divisible by 2`

### Steps to Reproduce

_No response_

### Important Factoids

_No response_

### References

_No response_

### LINKED ISSUE #20849: Existings `azurerm_app_configuration_key` broken after upgrade to 3.38.0 -> v3.46.0
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

v1.3.7

### AzureRM Provider Version

3.46.0

### Affected Resource(s)/Data Source(s)

azurerm_app_configuration_key

### Terraform Configuration Files

```hcl
resource "azurerm_app_configuration_key" "SharedSentinel" {
  configuration_store_id = azurerm_app_configuration.config.id

  key          = "SharedSentinel"
  value        = "4"
  content_type = ""
}
```

### Debug Output/Panic Output

```shell
╷
│ Error: parsing "/subscriptions/474fd901-117f-4540-9459-d79c2c7e2591/resourceGroups/some-rg/providers/Microsoft.AppConfiguration/configurationStores/some-az-config/AppConfigurationKey/SharedSentinel/Label/\x00": unexpected segment "AppConfigurationKey/SharedSentinel/Label/\x00" present at the end of the URI (input "/subscriptions/474fd901-117f-4540-9459-d79c2c7e2591/resourceGroups/some-rg/providers/Microsoft.AppConfiguration/configurationStores/some-az-config/AppConfigurationKey/SharedSentinel/Label/\x00")
│
│   with azurerm_app_configuration_key.SharedSentinel,
│   on appconfig.tf line 12, in resource "azurerm_app_configuration_key" "SharedSentinel":
│   12: resource "azurerm_app_configuration_key" "SharedSentinel" {
│
│ parsing "/subscriptions/474fd901-117f-4540-9459-d79c2c7e2591/resourceGroups/some-rg/providers/Microsoft.AppConfiguration/configurationStores/some-az-config/AppConfigurationKey/SharedSentinel/Label/\x00":
│ unexpected segment "AppConfigurationKey/SharedSentinel/Label/\x00" present at the end of the URI (input
│ "/subscriptions/474fd901-117f-4540-9459-d79c2c7e2591/resourceGroups/some-rg/providers/Microsoft.AppConfiguration/configurationStores/some-az-config/AppConfigurationKey/SharedSentinel/Label/\x00")
╵
```

### Expected Behaviour

Should not display an error

### Actual Behaviour

_No response_

### Steps to Reproduce

Previously existing state with this resources

### Important Factoids

was working in v3.38.0

### References

_No response_

### PR DESCRIPTION
**Pull request by teowa**
Resolves #20038
Resolves #20849

## Context:
The error is because of using `/AppConfigurationKey/SAMPLE/Label` as [label](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/app_configuration_key#label), which is sourced from the App Configuration Key ID format in Azurerm: `/subscriptions/00000000-0000-0000-0000-000000000000/resourcegroups/resourceGroup1/providers/Microsoft.AppConfiguration/configurationStores/appConf1/AppConfigurationKey/{appConfKey1}/Label/{label1}`, currently we support `/` in the `{appConfKey1}` and `{label1}` segments and stores the `/` directly in the resource ID (e.g. `/AppConfigurationKey/key:name/test/Label/test:label/name"` is a ID with `key:name/test` as the key and `test:label/name` as the label). This can ease the use of [import](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/app_configuration_key#import) feature with straight forward resource ID, but has some limitations like using `/Label/` or `/AppConfigurationKey/` literal in label will cause some conflict in ID parsing. To copy with using `/Label/` in label property, two possible solutions are:

1.  Escape the configuration Key and the label (e.g. transform `/Label/` to `%2FLabel%2F`) and then stores to the ID, but this makes the resource ID not so straight forward. 
2. Keep the current logic to store the `/Label/` directly to ID and forbidden using `/Label/` literal in `label` property.

~~This PR implements the **second** of above, keeps the current ID format (store the `/` instead of escaped `%2F`), but returns the error message telling users not to use `/Label/` literal label property, and since current used regex is greedy, we can use the `/AppConfigurationKey/` in label. what do you think?~~

---
## Update
change resource ID to valid [data plane URL format](https://learn.microsoft.com/en-us/azure/azure-app-configuration/rest-api-key-value): 
- `azurerm_app_configuration_feature`
  - `https://appconfname1.azconfig.io/kv/.appconfig.featureflag%2FkeyName?label=%00` (for no label) 
  - `https://appconfname1.azconfig.io/kv/.appconfig.featureflag%2FkeyName?label=labelName` (for when a label is present)
- `azurerm_app_configuration_key`
  - `https://appconfname1.azconfig.io/kv/keyName?label=labelName?label=%00` (for no label) 
  - `https://appconfname1.azconfig.io/kv/keyName?label=labelName` (for when a label is present)

the keyName in URL path should be escaped with Golang [`url.PathEscape`](https://pkg.go.dev/net/url#PathEscape) method and the labelName in URL query is escaped with [`url.QueryEscape`](https://pkg.go.dev/net/url#QueryEscape) method.

## Test:
```
make acctests SERVICE="appconfiguration" TESTARGS="-parallel 20 -test.run=TestAccAppConfigurationKey_compli" TESTTIMEOUT='1440m'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/appconfiguration -parallel 20 -test.run=TestAccAppConfigurationKey_compli -timeout 1440m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccAppConfigurationKey_complicatedKeyLabel
=== PAUSE TestAccAppConfigurationKey_complicatedKeyLabel
=== CONT  TestAccAppConfigurationKey_complicatedKeyLabel
--- PASS: TestAccAppConfigurationKey_complicatedKeyLabel (356.43s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/appconfiguration      359.890s
```

### CODE REVIEW

**Review by katbyte — COMMENTED** [review]
@teowa - i'm not sure this is the correct solution, given if it IS valid to have label in the name we should support it. This might be something we need to write a custom id parser for that is a little smarter.

@tombuildsstuff any ideas/comments?

**Comment by tombuildsstuff** [comment]
@teowa thanks for this PR.

So whilst this approach solves this issue to some extent - I think the other question here is why are these Data Plane resources using a Resource Manager ID at all? Whilst some of the functionality for App Configuration _is_ available within the Resource Manager API - [the primary interface for configuration is the Data Plane API](https://learn.microsoft.com/en-us/azure/azure-app-configuration/rest-api-key-value) which comes with it's own set of Resource IDs.

Whilst it's possible to use Resource Manager IDs for these resources, since the validation rules are different between Data Plane and Resource Manager (insofar as Resource Manager requires Key-Value pairs in the URI, whereas Data Plane allows any arbitrary expression) - so whilst URI-encoding the key would work - I think we'd be better to switch the Resource ID being used for these resources to match what the API's expecting here.

Doing so would allow the use of a more bespoke parser as @katbyte has suggested, which I'd agree would remove the problem here entirely - so I think would be the better approach here, and whilst this would require a State Migration to do, I think would probably make more sense than trying to shoe-horn this into the Resource Manager ID, since this is [ultimately using the Data Plane SDK/API](https://github.com/hashicorp/terraform-provider-azurerm/blob/036a8876978312bed45098aa52ad4b7023ce784d/internal/services/appconfiguration/app_configuration_key_resource.go#L127-L141).

As such would you be able to take a look into updating the Resource ID format being used for these resources to match:

1. `azurerm_app_configuration_feature` - `https://{store-name}.{dns-suffix-from-azure-environment}/keys?name=.appconfig.featureflag/{featureName}` (for no label) and `https://{store-name}.{dns-suffix-from-azure-environment/keys?name=.appconfig.featureflag/{featureName}&label={label}` (for when a label is present)
2. `azurerm_app_configuration_key` - `https://{store-name}.{dns-suffix-from-azure-environment}/keys?name={key}` (for no label) and `https://{store-name}.{dns-suffix-from-azure-environment}/keys?name={key}&label={label}` (for when a label is present)

Whilst that may seem like an odd choice - particularly for `azurerm_app_configuration_feature` - it both matches [how the API behaves](https://learn.microsoft.com/en-us/azure/azure-app-configuration/manage-feature-flags) and means that we can use Go's URI parsing logic to handle that - meaning we should be able to use the same approach as we do for the [Key Vault Nested Item parsers](https://github.com/hashicorp/terraform-provider-azurerm/blob/036a8876978312bed45098aa52ad4b7023ce784d/internal/services/keyvault/parse/nested_item.go#L92), rather than the generic `ParseResourceID` function which is intended for Azure Resource IDs.

As such, would you mind taking a look into adding the ID parsers, and then adding state migrations to update these?

Thanks!

**Comment by teowa** [comment]
Current parser accept either of below format ID for no label:
1. `https://{store-name}.azconfig.io/kv/{key}?label=`
2. `https://{store-name}.azconfig.io/kv/{key}?label=%00`
3. `https://{store-name}.azconfig.io/kv/{key}`

**Comment by katbyte** [comment]
@teowa - i think 1 or 2 is likely the best option, doesn't seem to be any point in making it more complex for a user by requiring `%00`. if i was a user i'd expect `https://{store-name}.azconfig.io/kv/{key}?label=` to give me an empty label.

could we also please update the PR title to reflect the id change? thanks

**Comment by teowa** [comment]
@katbyte The parser accept all of above format, but to keep consistence with [current ID format](https://github.com/hashicorp/terraform-provider-azurerm/pull/20082/files#diff-2ea23360bb5cdabcac3d9b33c7c952a8ce346e671ffc98679390201e988f6612L169-L172), `%00` is dumped to state for empty label.

**Comment by katbyte** [comment]
@teowa - the the parser should only accept 1 format to prevent confusion with users in what format to use and potentially issues later on - lets go with #1 as expecting users to know/handle %00 isn't useful?

File: website/docs/r/app_configuration_key.html.markdown [hunk 1]
```diff
@@ -163,11 +165,11 @@ The `timeouts` block allows you to specify [timeouts](https://www.terraform.io/l
 App Configuration Keys can be imported using the `resource id`, e.g.
 
 ```shell
-terraform import azurerm_app_configuration_key.test /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/resourceGroup1/providers/Microsoft.AppConfiguration/configurationStores/appConf1/AppConfigurationKey/appConfKey1/Label/label1
+terraform import azurerm_app_configuration_key.test https://appconfname1.azconfig.io/kv/keyName?label=labelName
 ```
 
-If you wish to import a key with an empty label then sustitute the label's name with `%00`, like this:
+If you wish to import a key with an empty label then substitute the label's name with empty, like this:
```
Comment (katbyte) [inline website/docs/r/app_configuration_key.html.markdown:171]: grammer
```suggestion
If you wish to import a key with an empty label then simply leave label's name blank:
```

File: website/docs/r/app_configuration_feature.html.markdown [hunk 2]
```diff
@@ -106,11 +106,11 @@ The `timeouts` block allows you to specify [timeouts](https://www.terraform.io/l
 App Configuration Features can be imported using the `resource id`, e.g.
 
 ```shell
-terraform import azurerm_app_configuration_feature.test /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/resourceGroup1/providers/Microsoft.AppConfiguration/configurationStores/appConf1/AppConfigurationFeature/appConfFeature1/Label/label1
+terraform import azurerm_app_configuration_feature.test https://appconfname1.azconfig.io/kv/.appconfig.featureflag%2FkeyName?label=labelName
 ```
 
-If you wish to import a key with an empty label then sustitute the label's name with `%00`, like this:
+If you wish to import a key with an empty label then substitute the label's name with empty, like this:
```
Comment (katbyte) [inline website/docs/r/app_configuration_feature.html.markdown:112]: as below

**Review by katbyte — APPROVED** [review]
aside from 2 docs comments this LGTM 🧪

**Comment by teowa** [comment]
Hi @katbyte, thanks for reviewing this. I have also enhanced the state migration to fix cases without label in issue https://github.com/hashicorp/terraform-provider-azurerm/issues/20849, please kindly take another look.

**Pull request merged**